README.md
# Ext-DateTime

[![LICENSE](https://img.shields.io/badge/release-0%2E0%2E0-blue.svg?style=flat)](hhttps://github.com/prokki/ext-datetime/releases/tag/0.0.0)
[![Packagist](https://img.shields.io/badge/Packagist-0%2E0%2E0-blue.svg?style=flat)](https://packagist.org/packages/prokki/ext-datetime)
[![LICENSE](https://img.shields.io/badge/License-MIT-blue.svg?style=flat)](LICENSE)
[![PHP v7.3](https://img.shields.io/badge/PHP-%E2%89%A57%2E3-0044aa.svg)](https://www.php.net/manual/en/migration73.new-features.php)
[![codecov](https://codecov.io/gh/prokki/ext-datetime/branch/master/graph/badge.svg)](https://codecov.io/gh/prokki/ext-datetime)
[![Build Status](https://travis-ci.org/prokki/ext-datetime.svg?branch=master)](https://travis-ci.org/prokki/ext-datetime)

Extends the native datetime objects ([\DateTime](https://www.php.net/manual/en/class.datetime.php) and [\DateTimeImmutable](https://www.php.net/manual/en/class.datetimeimmutable.php)) with a lot of additional helpful methods.

Most of the new methods are **short cuts of already existing functionality**.
But these methods will help you to save time and code if you need to handle
a many date- and time-operations.

All new methods support [*method chaining*](https://stackoverflow.com/questions/3724112/php-method-chaining). 

## Table of Contents

* [Requirements](#requirements)
* [Integration](#integration)
* [Usage](#usage)
  * [Static Initialization](#static-initialization)
  * [Cloning](#cloning)
  * [Manipulating](#manipulating)
  * [Instant Setter Methods](#instant-setter-methods)



## Requirements

The usage of [**PHP v7.3**](https://www.php.net/manual/en/migration73.new-features.php) is obligatory.



## Integration

Please install via [composer](https://getcomposer.org/).

```bash
composer require prokki/ext-datetime "^0.0"
```



## Usage

The classes and instances of the classes [ExtDateTime/DateTime](src/DateTime.php) and [ExtDateTime/DateTimeImmutable](src/DateTimeImmutable.php) can be used
exactly like native datetime objects/classes.

Example

```php
use ExtDateTime\DateTime;
use ExtDateTime\DateTimeImmutable;

// create a datetime object
$dateTime = new DateTime("now");

// create an immutable datetime object
$dateTimeImmutable = new DateTimeImmutable("now");
```


### Static Initialization

Similar to the native objects there are several others ways to create a datetime object.

All static methods can be used to implement [*method chaining*](https://stackoverflow.com/questions/3724112/php-method-chaining). 

#### create()

Use the static constructor `create()` to use chaining immediately.

```php
use ExtDateTime\DateTime;

// create a datetime object
// and use chaining immediately
$dateTime = DateTime::create("now")
            ->addHours(5)
            ->addDays(5)
            ;
```

#### current()

The static constructor `current()` returns a datetime object with the current date/time.
But in opposite to `DateTime::create("now")` this method returns the object **with additional microtime**. 

```php
use ExtDateTime\DateTime;

$currentMicroseconds = DateTime::current()->format("u");         // output example 654321
$noMicroseconds      = DateTime::create("now")->format("u");     // output always 000000
```

#### createFromObject()

Creates a new object from any datetime object implementing the [DateTimeInterface](https://www.php.net/manual/en/class.datetimeinterface.php).

```php
use ExtDateTime\DateTimeImmutable;

// create an immutable datetime object from a native non-immutable object
$datetime = DateTimeImmutable::createFromObject(new \DateTime());
```


### Cloning

Two new methods add the availability to use chaining directly after cloning an object.

#### duplicate()

This method is just a wrapper function for `clone`:

```php
use ExtDateTime\DateTime;

// create a datetime object
$datetime = DateTime::current();

// clone the datetime object and proceed like usual with chaining
$clone = $datetime->duplicate()
         ->addHours(5)
         ->addDays(5)
         ;
```

#### toImmutable() / toMutable()

Instead of using static constructors ([DateTime::createFromImmutable](https://www.php.net/manual/en/datetime.createfromimmutable.php) or [DateTimeImmutable::createFromMutable](https://www.php.net/manual/en/datetimeimmutable.createfrommutable.php)) you can also use these new non-static methods.

```php
use ExtDateTime\DateTime;

// create a datetime object
$datetime = DateTime::current();

$clonedImmutable = $datetime->toImmutable();       // only available in class DateTime

$clonedMutable   = $clonedImmutable->toMutable();  // only available in class DateTimeImmutable
```



### Manipulating

Most of the new methods are short cuts to avoid re-initialization of necessary parameter in your code.

#### addHours() / subHours()

Adds or subtracts hours of a datetime object.

```php
use ExtDateTime\DateTime;

// create a datetime object and add 10 hours
$datetimeFuture = DateTime::create("2020-07-30 12:35:17")
                  ->addHours(10)
                  ->format("Y-m-d h:i:s");                   // "2020-07-30 22:35:17"

// create a datetime object and subtracts 10 hours
$datetimeFuture = DateTime::create("2020-07-30 12:35:17")
                  ->subHours(10)
                  ->format("Y-m-d h:i:s");                   // "2020-07-30 02:35:17"
```

#### addDays() / subHours()

Adds or subtracts days of a datetime object.

```php
use ExtDateTime\DateTime;

// create a datetime object and add 10 days
$datetimeFuture = DateTime::create("2020-07-30 12:35:17")
                  ->addDays(10)
                  ->format("Y-m-d h:i:s");                   // "2020-08-09 22:35:17"

// create a datetime object and subtracts 10 days
$datetimeFuture = DateTime::create("2020-07-30 12:35:17")
                  ->subDays(10)
                  ->format("Y-m-d h:i:s");                   // "2020-07-20 02:35:17"
```

#### addMonth() / subMonth()

Adds or subtracts months of a datetime object.

**Attention**: This methods behaves differently than the suspected background function
of adding days. If the current date is the last day of the month (31/30/29/28) and the target month
has less days then the current month, the day will be set to the last day of the target month.

Example: The datetime object is

> *2017-01-30 17:00:00*

and you want to add _1 month_, the result will be

> *2017-02-28 17:00:00*


```php
use ExtDateTime\DateTime;

// create a datetime object and add 10 months
$datetimeFuture = DateTime::create("2020-07-30 12:35:17")
                  ->addMonth(10)
                  ->format("Y-m-d h:i:s");                   // "2020-08-09 22:35:17"

// create a datetime object and subtracts 10 months
$datetimeFuture = DateTime::create("2020-07-30 12:35:17")
                  ->subMonth(10)
                  ->format("Y-m-d h:i:s");                   // "2020-07-20 02:35:17"
```



### Instant Setter Methods

#### toEndOfDay()

Sets the time to the end of the day (*23:59:59*).

```php
use ExtDateTime\DateTime;

// create a datetime object and sets the time to the end of the day
$datetimeFuture = DateTime::create("2020-07-30 12:35:17")
                  ->toEndOfDay()
                  ->format("Y-m-d h:i:s");                   // "2020-07-30 23:59:59"
```

#### toNoon()

Sets the time to noon (*12:00:00*).

```php
use ExtDateTime\DateTime;

// create a datetime object and sets the time to noon
$datetimeFuture = DateTime::create("2020-07-30 12:35:17")
                  ->toNoon()
                  ->format("Y-m-d h:i:s");                   // "2020-07-30 12:00:00"
```

#### toStartOfDay()

Sets the time to the start of the day  (*00:00:00*).

```php
use ExtDateTime\DateTime;

// create a datetime object and sets the time the start of the day
$datetimeFuture = DateTime::create("2020-07-30 12:35:17")
                  ->toStartOfDay()
                  ->format("Y-m-d h:i:s");                   // "2020-07-30 00:00:00"
```

#### toStartOfMonth()

Sets the date to the first day of the month and additionally the time to the start of the day (*00:00:00*).

```php
use ExtDateTime\DateTime;

// create a datetime object and sets the date to the first day of the month
$datetimeFuture = DateTime::create("2020-07-30 12:35:17")
                  ->toStartOfMonth()
                  ->format("Y-m-d h:i:s");                   // "2020-07-01 00:00:00"
```

#### toEndOfMonth()

Sets the date to the last day of the month and additionally the time to the end of the day (*23:59:59*).

```php
use ExtDateTime\DateTime;

// create a datetime object and sets the date to the last day of the month
$datetimeFuture = DateTime::create("2020-07-30 12:35:17")
                  ->toEndOfMonth()
                  ->format("Y-m-d h:i:s");                   // "2020-07-31 23:59:59"
```


CHANGELOG.md
# Changes in Ext-DateTime

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [0.0.0] - 2019-07-31

### Added
- First version supports
  - Static constructors
    - `DateTime::create()`
    - `DateTime::current()`
    - `DateTime::createFromObject()`
  - Non-static cloning
    - `DateTime::duplicate()`
    - `DateTime::toImmutable()` and `DateTimeImmutable::toMutable()` 
  - Manipulating methods
    - `DateTime::addHours()` / `DateTime::subHours()`
    - `DateTime::addDays()` / `DateTime::subDays()`
    - `DateTime::addMonth()` / `DateTime::subMonth()`
  - Instant setters
    - `DateTime::toEndOfDay()`, `DateTime::toNoon()`, `DateTime::toStartOfDay()`
    - `DateTime::toStartOfMonth()`, `DateTime::toEndOfMonth()`


tests/DateTimeTest.php
<?php

namespace ExtDateTime\Test;

use DateTimeZone;
use Exception;
use ExtDateTime\DateTime;
use PHPUnit\Framework\TestCase;

class DateTimeTest extends TestCase
{
    /**
     * {@inheritDoc}
     */
    public function setUp(): void
    {
        // call super method
        parent::setUp();

        // server timezone is UTC
        date_default_timezone_set('UTC');
    }

    /**
     * @throws Exception
     */
    public static function testCurrent()
    {
        $dateTime1 = DateTime::current();
        $dateTime2 = DateTime::current();
        self::assertNotEquals($dateTime1, $dateTime2);

        $today = new \DateTime('now');
        self::assertLessThan(1000000, abs((int) $today->diff($dateTime1)->format('%f')));
    }

    /**
     * @throws Exception
     */
    public static function testCreateFromObject()
    {
        $dateTime1 = DateTime::createFromObject(new \DateTime('2016-07-13T19:25:47+07:00'));
        $dateTime2 = DateTime::createFromFormat('Y-m-d H:i:s', '2016-07-13 19:25:47', new DateTimeZone('+07:00'));

        self::assertEquals($dateTime1, $dateTime2);
    }

    /**
     * @throws Exception
     */
    public function testCreateFromFormat()
    {
        $dateTime = DateTime::createFromFormat('Y-m-d H:i:s', '2019-12-31 16:17:35');

        self::assertInstanceOf(DateTime::class, $dateTime);
        self::assertEquals(1577809055, $dateTime->getTimestamp());
    }

    /**
     * @throws Exception
     */
    public function testCreateFromFormatException()
    {
        $this->expectException(Exception::class);
        $this->expectExceptionMessage('Cannot create an object by DateTime::createFromFormat().');

        DateTime::createFromFormat('Y-m-d', 'I do not want to submit valid parameters!');
    }

    /**
     * @throws Exception
     */
    public function testCreateFromImmutable()
    {
        $dateTime = DateTime::createFromImmutable(new \DateTimeImmutable());
        $dateTimeChanged = $dateTime->addDays(1);

        self::assertNotInstanceOf(\DateTimeImmutable::class, $dateTime);
        self::assertSame($dateTime, $dateTimeChanged);
    }

    /**
     * @throws Exception
     */
    public function testCreate()
    {
        self::assertInstanceOf(DateTime::class, DateTime::create());
    }

    /**
     * @throws Exception
     */
    public function testToImmutable()
    {
        $dateTime = DateTime::create();
        $dateTimeImmutable = $dateTime->toImmutable();
        $dateTimeImmutableChanged = $dateTimeImmutable->addDays(1);

        self::assertInstanceOf(\DateTimeImmutable::class, $dateTimeImmutable);
        self::assertInstanceOf(\DateTimeImmutable::class, $dateTimeImmutableChanged);
        self::assertNotSame($dateTimeImmutable, $dateTimeImmutableChanged);
    }

    /**
     * @throws Exception
     */
    public function testDuplicate()
    {
        $dateTime1 = DateTime::create();
        $dateTime2 = $dateTime1->duplicate();

        // content of date time equals
        self::assertEquals($dateTime1->format('U.u'), $dateTime2->format('U.u'));
        // but objects are not the same
        self::assertNotSame($dateTime1, $dateTime2);
        self::assertInstanceOf(DateTime::class, $dateTime2);
    }

    /**
     * @throws Exception
     */
    public function testAddHours()
    {
        $dateTime = DateTime::create('2019-12-31 16:17:35');

        // check "same" object
        self::assertSame($dateTime, $dateTime->addHours(0));
        self::assertInstanceOf(DateTime::class, $dateTime->addHours(0));

        // check +8 hours from origin $dateTime
        self::assertEquals('2020-01-01 00:17:35', $dateTime->addHours(8)->format('Y-m-d H:i:s'));
        // check -46 hours from already changed $dateTime
        self::assertEquals('2019-12-30 02:17:35', $dateTime->addHours(-46)->format('Y-m-d H:i:s'));
    }

    /**
     * @throws Exception
     */
    public function testSubHours()
    {
        $dateTime = DateTime::create('1978-08-20 07:14:57');

        // check "same" object
        self::assertSame($dateTime, $dateTime->subHours(0));
        self::assertInstanceOf(DateTime::class, $dateTime->subHours(0));

        // check -100 hours from origin $dateTime
        self::assertEquals('1978-08-16 03:14:57', $dateTime->subHours(100)->format('Y-m-d H:i:s'));
    }

    /**
     * @throws Exception
     */
    public function testAddDays()
    {
        $dateTime = DateTime::create('2019-12-31 16:17:35');

        // check "same" object
        self::assertSame($dateTime, $dateTime->addDays(0));
        self::assertInstanceOf(DateTime::class, $dateTime->addDays(0));

        // check +8 hours from origin $dateTime
        self::assertEquals('2020-01-08 16:17:35', $dateTime->addDays(8)->format('Y-m-d H:i:s'));
        // check -46 hours from already changed $dateTime
        self::assertEquals('2019-11-23 16:17:35', $dateTime->addDays(-46)->format('Y-m-d H:i:s'));
    }

    /**
     * @throws Exception
     */
    public function testSubDays()
    {
        $dateTime = DateTime::create('1978-08-20 07:14:57');

        // check "same" object
        self::assertSame($dateTime, $dateTime->subDays(0));
        self::assertInstanceOf(DateTime::class, $dateTime->subDays(0));

        // check -100 days from origin $dateTime
        self::assertEquals('1978-05-12 07:14:57', $dateTime->subDays(100)->format('Y-m-d H:i:s'));
    }

    /**
     * @throws Exception
     */
    public function testAddMonth()
    {
        $dateTime = DateTime::create('2019-12-31 16:17:35');

        // check "same" object
        self::assertSame($dateTime, $dateTime->addMonth(0));
        self::assertInstanceOf(DateTime::class, $dateTime->addMonth(0));

        // check +8 months from origin $dateTime
        self::assertEquals('2020-08-31 16:17:35', $dateTime->addMonth(8)->format('Y-m-d H:i:s'));
        // check -46 months from already changed $dateTime
        self::assertEquals('2016-10-31 16:17:35', $dateTime->addMonth(-46)->format('Y-m-d H:i:s'));

        // special case: add one month to the 31th of October = not November, 31st but November, 30th!
        self::assertEquals('2016-11-30 16:17:35', DateTime::create('2016-10-31 16:17:35')->addMonth(1)->format('Y-m-d H:i:s'));
    }

    /**
     * @throws Exception
     */
    public function testSubMonth()
    {
        $dateTime = DateTime::create('2019-12-31 16:17:35');

        // check "same" object
        self::assertSame($dateTime, $dateTime->subMonth(0));
        self::assertInstanceOf(DateTime::class, $dateTime->subMonth(0));

        // check -100 months from origin $dateTime
        self::assertEquals('2011-08-31 16:17:35', $dateTime->subMonth(100)->format('Y-m-d H:i:s'));

        // special case: sub two month from the 31th of August = not June, 31st but June, 30th!
        self::assertEquals('2011-06-30 16:17:35', DateTime::create('2011-08-31 16:17:35')->subMonth(2)->format('Y-m-d H:i:s'));
    }

    /**
     * @throws Exception
     */
    public function testToEndOfDay()
    {
        $dateTime = DateTime::create('1978-08-20 07:14:57');

        // check "same" object
        self::assertSame($dateTime, $dateTime->toEndOfDay());
        self::assertInstanceOf(DateTime::class, $dateTime->toEndOfDay());

        // set time to end of day (23:59:59)
        self::assertEquals('1978-08-20 23:59:59', $dateTime->toEndOfDay()->format('Y-m-d H:i:s'));
    }

    /**
     * @throws Exception
     */
    public function testToNoon()
    {
        $dateTime = DateTime::create('1978-08-20 07:14:57');

        // check "same" object
        self::assertSame($dateTime, $dateTime->toNoon());
        self::assertInstanceOf(DateTime::class, $dateTime->toNoon());

        // set time to noon (12:00:00)
        self::assertEquals('1978-08-20 12:00:00', $dateTime->toNoon()->format('Y-m-d H:i:s'));
    }

    /**
     * @throws Exception
     */
    public function testToStartOfDay()
    {
        $dateTime = DateTime::create('1978-08-20 07:14:57');

        // check "same" object
        self::assertSame($dateTime, $dateTime->toStartOfDay());
        self::assertInstanceOf(DateTime::class, $dateTime->toStartOfDay());

        // set time to start of day (00:00:00)
        self::assertEquals('1978-08-20 00:00:00', $dateTime->toStartOfDay()->format('Y-m-d H:i:s'));
    }

    /**
     * @throws Exception
     */
    public function testToStartOfMonth()
    {
        $dateTime = DateTime::create('1978-08-20 07:14:57');

        // check "same" object
        self::assertSame($dateTime, $dateTime->toStartOfMonth());
        self::assertInstanceOf(DateTime::class, $dateTime->toStartOfMonth());

        // sets the date to the end of the month (last day of the month) and the time to end of the day (23:59:59)
        self::assertEquals('1978-08-01 00:00:00', $dateTime->toStartOfMonth()->format('Y-m-d H:i:s'));
        self::assertEquals('1978-08-01 00:00:00', $dateTime->toStartOfMonth()->format('Y-m-d H:i:s'));
    }

    /**
     * @throws Exception
     */
    public function testToEndOfMonth()
    {
        $dateTime = DateTime::create('1978-08-20 07:14:57');

        // check "same" object
        self::assertSame($dateTime, $dateTime->toEndOfMonth());
        self::assertInstanceOf(DateTime::class, $dateTime->toEndOfMonth());

        // sets the date to the start of the month (first day of the month) and the time to start of the day (00:00:00)
        self::assertEquals('1978-08-31 23:59:59', $dateTime->toEndOfMonth()->format('Y-m-d H:i:s'));
        self::assertEquals('1978-08-31 23:59:59', $dateTime->toEndOfMonth()->format('Y-m-d H:i:s'));
    }

}


src/DateTime.php
<?php

declare( strict_types=1 );

namespace ExtDateTime;

use DateInterval;
use DateTimeInterface;
use DateTimeZone;
use Exception;

/**
 * Class DateTime
 *
 * @package ExtDateTime
 */
class DateTime extends \DateTime
{
    /**
     * Returns the current server date time or `false` if the current day could not be fetched.
     *
     * @return boolean|static
     *
     * @throws Exception
     */
    public static function current()
    {
        return static::createFromFormat('U.u', sprintf('%.6F', microtime(true)));
    }

    /**
     * Parses a string into a new {@see DateTime} object according to the specified format.
     *
     * @param string            $format   Format accepted by date().
     * @param string            $time     String representing the time.
     * @param DateTimeZone|null $timezone [optional] A DateTimeZone object representing the desired time zone.
     *
     * @return boolean|static
     *
     * @throws Exception
     *
     * @link http://php.net/manual/en/datetime.createfromformat.php
     */
    public static function createFromFormat($format, $time, DateTimeZone $timezone = null)
    {
        $datetime = is_null($timezone) ?
            parent::createFromFormat($format, $time) :
            parent::createFromFormat($format, $time, $timezone);

        if( !is_object($datetime) )
        {
            throw new Exception('Cannot create an object by DateTime::createFromFormat().');
        }

        return static::createFromObject($datetime);
    }

    /**
     * Overrides {@see \DateTime::createFromImmutable()} to return a {@see \ExtDateTime\DateTime} object
     * instead of the default {@see \DateTime} object.
     *
     * @param \DateTimeImmutable $DateTimeImmutable
     *
     * @return boolean|static
     *
     * @throws Exception
     */
    public static function createFromImmutable($DateTimeImmutable)
    {
        return static::createFromObject($DateTimeImmutable);
    }

    /**
     * Copies and casts the submitted datetime object to a new {@see \ExtDateTime\DateTime} object.
     *
     * @param DateTimeInterface $dateTime the source object has to implement the {@see \DateTimeInterface}
     *
     * @return boolean|static
     */
    public static function createFromObject(DateTimeInterface $dateTime)
    {
        $parts      = explode(':', serialize($dateTime));
        $parts[ 1 ] = strlen(static::class);
        $parts[ 2 ] = sprintf('"%s"', static::class);

        return unserialize(implode(':', $parts));
    }

    /**
     * Static constructor. Returns a new {@see \ExtDateTime\DateTime} object.
     *
     * @param string       $time
     * @param DateTimeZone $timezone
     *
     * @return boolean|static
     *
     * @throws Exception
     */
    public static function create(string $time = 'now', DateTimeZone $timezone = null)
    {
        return new static($time, $timezone);
    }

    /**
     * Converts the mutable datetime object to an immutable datetime object.
     *
     * Wrapper function of {\ExtDateTime\DateTimeImmutable::createFromMutable()}.
     *
     * @return DateTimeImmutable
     *
     * @throws Exception
     */
    public function toImmutable(): DateTimeImmutable
    {
        return DateTimeImmutable::createFromMutable($this);
    }

    /**
     * Returns a cloned object.
     *
     * This method is just a wrapper function for {@see \clone()}.
     *
     * @return static
     */
    public function duplicate(): DateTimeInterface
    {
        return clone $this;
    }

    /**
     * @return integer
     */
    protected function getAbsoluteMonths(): int
    {
        return intval($this->format('Y')) * 12 + intval($this->format('m'));
    }

    /**
     * Adds (or subtracts) hours.
     *
     * @param integer $hours to add some hours pass a positive number, to subtract pass a negative number
     *
     * @return static
     */
    public function addHours(int $hours): DateTimeInterface
    {
        return $this->add(DateInterval::createFromDateString(sprintf('%d hour', $hours)));
    }

    /**
     * Subtracts (or adds) hours.
     *
     * @param integer $hours to subtract some hours pass a positive number, to add pass a negative number
     *
     * @return static
     */
    public function subHours(int $hours): DateTimeInterface
    {
        return $this->addHours(0 - $hours);
    }

    /**
     * Adds (or subtracts) days.
     *
     * @param integer $days to add some days pass a positive number, to subtract pass a negative number
     *
     * @return static
     */
    public function addDays(int $days): DateTimeInterface
    {
        return $this->add(DateInterval::createFromDateString(sprintf('%d day', $days)));
    }

    /**
     * Subtracts (or adds) days.
     *
     * @param integer $days to subtracts some days pass a positive number, to add pass a negative number
     *
     * @return static
     */
    public function subDays(int $days): DateTimeInterface
    {
        return $this->addDays(0 - $days);
    }

    /**
     * Adds an amount of months to the current date.
     *
     * **ATTENTION**: If the current date is the 31/30/29 and the target month has less days then the current month, the day will be set to the last
     * day of the target month.
     *
     * Example: Current date is '2017-01-30 17:00:00' and you want to add '1 month', then the result will be '2017-02-28 17:00:00'.
     *
     * @param integer $month
     *
     * @return static
     */
    public function addMonth(int $month): DateTimeInterface
    {
        $absoluteMonths = $this->getAbsoluteMonths();

        $this->add(DateInterval::createFromDateString(sprintf('%d month', $month)));

        if( $absoluteMonths + $month !== $this->getAbsoluteMonths() )
        {
            $this->subDays((int) $this->format('d'));
        }

        return $this;
    }

    /**
     * Subs an amount of months from the current date.
     *
     * **ATTENTION**: If the current date is the 31/30/29 and the target month has less days then the current month, the day will be set to the last
     * day of the target month.
     *
     * Example: Current date is '2017-03-31 17:00:00' and you want to sub '1 month', then the result will be '2017-02-28 17:00:00'.
     *
     * @param integer $month
     *
     * @return static
     */
    public function subMonth(int $month): DateTimeInterface
    {
        $absolute_months = $this->getAbsoluteMonths();

        $this->sub(DateInterval::createFromDateString(sprintf('%d month', $month)));

        if( $absolute_months - $month !== $this->getAbsoluteMonths() )
        {
            $this->subDays((int) $this->format('d'));
        }

        return $this;
    }

    /**
     * Sets the time to the end of the day (`23:59:59`).
     *
     * @return static
     */
    public function toEndOfDay(): DateTimeInterface
    {
        return $this->setTime(23, 59, 59);
    }

    /**
     * Sets the time to noon (`12:00:00`).
     *
     * @return static
     */
    public function toNoon(): DateTimeInterface
    {
        return $this->setTime(12, 0, 0);
    }

    /**
     * Sets the time to the start of the day (`00:00:00`).
     *
     * @return static
     */
    public function toStartOfDay(): DateTimeInterface
    {
        return $this->setTime(0, 0, 0);
    }

    /**
     * Sets the date to the first day of the month and additionally the time to the start of the day (`00:00:00`).
     *
     * @return static
     */
    public function toStartOfMonth(): DateTimeInterface
    {
        return $this->setDate((int) $this->format('Y'), (int) $this->format('m'), 1)->toStartOfDay();
    }

    /**
     * Sets the date to the last day of the month and additionally the time to the end of the day (`23:59:59`).
     *
     * @return static
     */
    public function toEndOfMonth(): DateTimeInterface
    {
        return $this->setDate((int) $this->format('Y'), (int) $this->format('m'), (int) $this->format('t'))->toEndOfDay();
    }

}


src/DateTimeImmutable.php
<?php

declare( strict_types=1 );

namespace ExtDateTime;

use DateInterval;
use DateTimeInterface;
use DateTimeZone;
use Exception;

/**
 * Class DateTimeImmutable
 *
 * @package ExtDateTime
 */
class DateTimeImmutable extends \DateTimeImmutable
{
    /**
     * Returns the current server date time or `false` if the current day could not be fetched.
     *
     * @return boolean|static
     *
     * @throws Exception
     */
    public static function current()
    {
        return static::createFromFormat('U.u', sprintf('%.6F', microtime(true)));
    }

    /**
     * Parses a string into a new DateTime object according to the specified format.
     *
     * @param string            $format   Format accepted by date().
     * @param string            $time     String representing the time.
     * @param DateTimeZone|null $timezone [optional] A DateTimeZone object representing the desired time zone.
     *
     * @return static|boolean
     *
     * @throws Exception
     *
     * @link http://php.net/manual/en/datetime.createfromformat.php
     */
    public static function createFromFormat($format, $time, DateTimeZone $timezone = null)
    {
        $datetime = is_null($timezone) ?
            parent::createFromFormat($format, $time) :
            parent::createFromFormat($format, $time, $timezone);

        if( !is_object($datetime) )
        {
            throw new Exception('Cannot create an object by DateTime::createFromFormat().');
        }

        return static::createFromObject($datetime);
    }

    /**
     * Overrides {@see \DateTime::createFromMutable()} to return a {@see \ExtDateTime\DateTimeImmutable} object
     * instead of the default {@see \DateTimeImmutable} object.
     *
     * @param \DateTime $DateTimeMutable
     *
     * @return boolean|static
     *
     * @throws Exception
     */
    public static function createFromMutable($DateTimeMutable)
    {
        return static::createFromObject($DateTimeMutable);
    }

    /**
     * Copies and casts the submitted datetime object to a new {@see \ExtDateTime\DateTimeImmutable} object.
     *
     * @param DateTimeInterface $dateTime the source object has to implement the {@see \DateTimeInterface}
     *
     * @return boolean|static
     *
     * @throws Exception
     */
    public static function createFromObject(DateTimeInterface $dateTime)
    {
        $parts      = explode(':', serialize($dateTime));
        $parts[ 1 ] = strlen(static::class);
        $parts[ 2 ] = sprintf('"%s"', static::class);

        return unserialize(implode(':', $parts));
    }

    /**
     * Static constructor. Returns a new {@see \ExtDateTime\DateTimeImmutable} object.
     *
     * @param string       $time
     * @param DateTimeZone $timezone
     *
     * @return boolean|static
     *
     * @throws Exception
     */
    public static function create(string $time = 'now', DateTimeZone $timezone = null)
    {
        return new static($time, $timezone);
    }

    /**
     * Converts the immutable datetime object to a mutable datetime object.
     *
     * Wrapper function of {\ExtDateTime\DateTime::createFromImmutable()}.
     *
     * @return DateTime
     *
     * @throws Exception
     */
    public function toMutable(): DateTime
    {
        return DateTime::createFromImmutable($this);
    }

    /**
     * Returns a cloned object.
     *
     * This method is just a wrapper function for {@see \clone()}.
     *
     * @return static
     */
    public function duplicate(): DateTimeInterface
    {
        return clone $this;
    }

    /**
     * @return integer
     */
    protected function getAbsoluteMonths(): int
    {
        return intval($this->format('Y')) * 12 + intval($this->format('m'));
    }

    /**
     * Adds (or subtracts) hours.
     *
     * @param integer $hours to add some hours pass a positive number, to subtract pass a negative number
     *
     * @return static
     */
    public function addHours(int $hours): DateTimeInterface
    {
        return $this->add(DateInterval::createFromDateString(sprintf('%d hour', $hours)));
    }

    /**
     * Subtracts (or adds) hours.
     *
     * @param integer $hours to subtract some hours pass a positive number, to add pass a negative number
     *
     * @return static
     */
    public function subHours(int $hours): DateTimeInterface
    {
        return $this->addHours(0 - $hours);
    }

    /**
     * Adds (or subtracts) days.
     *
     * @param integer $days to add some days pass a positive number, to subtract pass a negative number
     *
     * @return static
     */
    public function addDays(int $days): DateTimeInterface
    {
        return $this->add(DateInterval::createFromDateString(sprintf('%d day', $days)));
    }

    /**
     * Subtracts (or adds) days.
     *
     * @param integer $days to subtracts some days pass a positive number, to add pass a negative number
     *
     * @return static
     */
    public function subDays(int $days): DateTimeInterface
    {
        return $this->addDays(0 - $days);
    }

    /**
     * Adds an amount of months to the current date.
     *
     * **ATTENTION**: If the current date is the 31/30/29 and the target month has less days then the current month, the day will be set to the last
     * day of the target month.
     *
     * Example: Current date is '2017-01-30 17:00:00' and you want to add '1 month', then the result will be '2017-02-28 17:00:00'.
     *
     * @param integer $month
     *
     * @return static
     */
    public function addMonth(int $month): DateTimeInterface
    {
        $absoluteMonths = $this->getAbsoluteMonths();

        $dateTimeImmutable = $this->add(DateInterval::createFromDateString(sprintf('%d month', $month)));

        if( $absoluteMonths + $month !== $dateTimeImmutable->getAbsoluteMonths() )
        {
            $dateTimeImmutable = $dateTimeImmutable->subDays((int) $dateTimeImmutable->format('d'));
        }

        return $dateTimeImmutable;
    }

    /**
     * Subs an amount of months from the current date.
     *
     * **ATTENTION**: If the current date is the 31/30/29 and the target month has less days then the current month, the day will be set to the last
     * day of the target month.
     *
     * Example: Current date is '2017-03-31 17:00:00' and you want to sub '1 month', then the result will be '2017-02-28 17:00:00'.
     *
     * @param integer $month
     *
     * @return static
     */
    public function subMonth(int $month): DateTimeInterface
    {
        $absoluteMonths = $this->getAbsoluteMonths();

        $dateTimeImmutable = $this->sub(DateInterval::createFromDateString(sprintf('%d month', $month)));

        if( $absoluteMonths - $month !== $dateTimeImmutable->getAbsoluteMonths() )
        {
            $dateTimeImmutable = $dateTimeImmutable->subDays((int) $dateTimeImmutable->format('d'));
        }

        return $dateTimeImmutable;
    }

    /**
     * Sets the time to the end of the day (`23:59:59`).
     *
     * @return static
     */
    public function toEndOfDay(): DateTimeInterface
    {
        return $this->setTime(23, 59, 59);
    }

    /**
     * Sets the time to noon (`12:00:00`).
     *
     * @return static
     */
    public function toNoon(): DateTimeInterface
    {
        return $this->setTime(12, 0, 0);
    }

    /**
     * Sets the time to the start of the day (`00:00:00`).
     *
     * @return static
     */
    public function toStartOfDay(): DateTimeInterface
    {
        return $this->setTime(0, 0, 0);
    }

    /**
     * Sets the date to the first day of the month and additionally the time to the start of the day (`00:00:00`).
     *
     * @return static
     */
    public function toStartOfMonth(): DateTimeInterface
    {
        return $this->setDate((int) $this->format('Y'), (int) $this->format('m'), 1)->toStartOfDay();
    }

    /**
     * Sets the date to the last day of the month and additionally the time to the end of the day (`23:59:59`).
     *
     * @return static
     */
    public function toEndOfMonth(): DateTimeInterface
    {
        return $this->setDate((int) $this->format('Y'), (int) $this->format('m'), (int) $this->format('t'))->toEndOfDay();
    }

}
